go/database/mpt/tool/stress.go

```diff
@@ -0,0 +1,348 @@
+		// deleting reasonable large accounts with a few dozen
+		// slots.
+		var err error
+		switch c := rand.Float32(); {
+		case c < 0.65:
+			err = s.addSlot()
+		case c < 0.995:
+			err = s.updateSlot()
+		default:
+			err = s.deleteAccount()
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	if _, _, err := s.db.UpdateHashes(); err != nil {
+		return fmt.Errorf("failed to update hashes: %w", err)
+	}
+
+	s.blockHeight++
+	return nil
+}
+
+func (s *stressTestState) addSlot() error {
+	isNewAccount := false
+	addrIndex := 0
+	if len(s.state) > 0 && rand.Float32() < 0.98 { // < most of the time an old account is re-used
+		addrIndex = s.getRandomAccountIndex()
+	} else {
+		addrIndex = s.nextAccount
+		s.nextAccount++
+		isNewAccount = true
+	}
+	addr := intToAddress(addrIndex)
+
+	if isNewAccount {
+		s.state[addrIndex] = map[int]int{}
+		if err := s.db.SetNonce(addr, common.ToNonce(1)); err != nil {
+			return fmt.Errorf("failed to create account: %w", err)
+		}
+	}
+
+	storage := s.state[addrIndex]
+	keyIndex := s.nextKey
+	s.nextKey++
+	key := intToKey(keyIndex)
+
+	current, err := s.db.GetStorage(addr, key)
+	if err != nil {
+		return fmt.Errorf("failed to get value: %w", err)
+	}
+	if want, got := (common.Value{}), current; want != got {
+		return fmt.Errorf("unexpected value %d/%d - wanted %x, got %x", addrIndex, keyIndex, want, got)
+	}
+
+	value := intToValue(1)
+	if err := s.db.SetStorage(addr, key, value); err != nil {
+		return fmt.Errorf("failed to set value: %w", err)
+	}
+	storage[keyIndex] = 1
+	return nil
+}
+
+func (s *stressTestState) updateSlot() error {
+	if len(s.state) == 0 {
+		return nil
+	}
+	addrIndex := s.getRandomAccountIndex()
+	addr := intToAddress(addrIndex)
+	storage := s.state[addrIndex]
+
+	keyIndex := 0
+	for i := range storage {
+		keyIndex = i
+		break
+	}
+	key := intToKey(keyIndex)
+
+	current, err := s.db.GetStorage(addr, key)
+	if err != nil {
+		return fmt.Errorf("failed to get value: %w", err)
+	}
+	if want, got := intToValue(storage[keyIndex]), current; want != got {
+		return fmt.Errorf("unexpected value %v/%v before update - wanted %x, got %x", addr, key, want, got)
+	}
+
+	newValue := storage[keyIndex] + 1
+	value := intToValue(newValue)
+	if err := s.db.SetStorage(addr, key, value); err != nil {
+		return fmt.Errorf("failed to set value: %w", err)
+	}
+	storage[keyIndex] = newValue
+	return nil
+}
+
+func (s *stressTestState) deleteAccount() error {
+	if len(s.state) == 0 {
+		return nil
+	}
+	addrIndex := s.getRandomAccountIndex()
+	if false && len(s.state[addrIndex]) > 50 {
+		fmt.Printf("deleting %d with %d keys\n", addrIndex, len(s.state[addrIndex]))
+	}
+	addr := intToAddress(addrIndex)
+	if err := s.db.DeleteAccount(addr); err != nil {
+		return fmt.Errorf("failed to remove account: %w", err)
+	}
+	delete(s.state, addrIndex)
+	return nil
+}
+
+func (s *stressTestState) getRandomAccountIndex() int {
+	for i := range s.state { // iteration order is random, we pick the first one
```

Comment:
I think iteration order is random between go runtime runs, but in one run, when this method is called many times, the order will be always the same. I am not sure what was intention here.

Reply:
This is actually random per invocation. See [example](https://go.dev/play/p/biHhzXwXhjJ).

Reply:
I see, thanks for explanation